### PR TITLE
PBL-6583 File rename, PBL-17067 Allow users to update images in CloudPebble

### PR DESCRIPTION
**Pull request by Spacerat**
This PR adds the ability to rename code and resource files. It also fixes the following issues:
- PBL-17067 i.e. updating a resource with a file with the same name as its current name would fail
- Uploading a new variant for an existing PNG resource would not immediately show the new image
- CloudPebble now validates filenames

### CODE REVIEW

File: ide/models/files.py [hunk 1]
```diff
@@ -240,6 +251,13 @@ def get_contents(self):
         else:
             return s3.read_file('source', self.s3_path)
 
+    def was_modified_since(self, expected_modification_time):
+        if isinstance(expected_modification_time, int):
+            expected_modification_time = datetime.datetime.fromtimestamp(expected_modification_time)
+        else:
```
Comment (Katharine) [inline ide/models/files.py]: Why bother with the `else`? The assertion should hold either way.

File: ide/models/files.py [hunk 2]
```diff
@@ -36,6 +39,13 @@ def get_best_variant(self, variant):
         except ResourceVariant.DoesNotExist:
             return self.variants.get(variant=0)
 
+    def rename(self, new_name):
+        #if not re.match(r"^[\.a-zA-Z0-9-_ ]+\.[a-zA-Z0-9]+$", new_name):
```
Comment (Katharine) [inline ide/models/files.py]: Don't leave commented-out code.

File: ide/models/files.py (hunk 1 above)
Comment (Spacerat) [inline ide/models/files.py]: Ah good catch, it clearly shouldn't be in an else

File: ide/templates/ide/project/resource.html [hunk 3]
```diff
@@ -78,7 +84,7 @@
             </div>
         </div>
         <div class="control-group well native-only" id="non-font-resource-group">
-            <label class="control-label" for="edit-resource-id">{% trans 'Identifier' %}</label>
+            <label class="control-label" id="edit-resource-id">{% trans 'Identifier' %}</label>
```
Comment (Katharine) [inline ide/templates/ide/project/resource.html]: Why?

File: ide/static/ide/js/resources.js [hunk 4]
```diff
@@ -314,11 +324,28 @@ CloudPebble.Resources = (function() {
             var form = pane.find('form');
             form.submit(function(e) {
                 e.preventDefault();
-                process_resource_form(form, false, "/ide/project/" + PROJECT_ID + "/resource/" + resource.id + "/update", function(data) {
+                process_resource_form(form, false, resource.file_name, "/ide/project/" + PROJECT_ID + "/resource/" + resource.id + "/update", function(data) {
                     // Update any previews we have.
-                    $('.image-resource-preview img').each(function() {
-                        $(this).attr('src', $(this).attr('src').replace(/#e.*$/,'') + '#e' + (++preview_count));
+                    _.each(data.variants, function(variant) {
+                        var variant_div = $('.variant-'+variant);
```
Comment (Katharine) [inline ide/static/ide/js/resources.js]: You probably want `pane.find` here instead of `$`.

File: ide/static/ide/js/resources.js [hunk 5]
```diff
@@ -314,11 +324,28 @@ CloudPebble.Resources = (function() {
...
+                            if (!old_src || /^(\s*(\?t=.*)|\s+)$/.test(old_src)) {
+                                // If we're uploading a new variant, we need to set its preview url from scratch
+                                return '/ide/project/' + PROJECT_ID + '/resource/' + resource.id + '/' + variant + '/get';
+                            }
+                            else {
+                                // If not, we have to refresh and bypass the cache the image using a query string
+                                return old_src.replace(/\?t=.*$/,'') + '?t=' + (++preview_count);
+                            }
+                        });
+                        if (resource.kind == 'png' || resource.kind == 'png-trans') {
```
Comment (Katharine) [inline ide/static/ide/js/resources.js]: Is this the best place to make this check? If we will never show it, why are we bothering to try assembling things to put in it?

File: ide/templates/ide/project/resource.html (hunk 3 above)
Comment (Spacerat) [inline ide/templates/ide/project/resource.html]: Oops. Mind you, I guess I was probably half way through trying to fix a label for/id mismatch and then forgot. At the moment a lot of labels in resource.html aren't actually tied to anything, but maybe fixing them all should be done in another PR.

File: ide/static/ide/js/resources.js (hunk 5 above)
Comment (Spacerat) [inline ide/static/ide/js/resources.js]: Oh, quite right
